Add an edit-mode option

evertrol opened the pull request:
This adds an option to run from a local repository in edit mode, where
changes in a running Docker container (for example, through a
notebook) are reflected in the local repository.

Implements the feature suggested in #357.

evertrol commented:
This may still require some tests, which are probably quite a bit more work than implementing the actual feature (given that changes in both a local directory and a container need to be verified).

betatim commented:
Test will indeed require a bit of thought. Some ideas below.

What do you think of using the tests from #413 (https://github.com/jupyter/repo2docker/pull/413/files#diff-73b9b926dafec45bd12128b598a0dd45) as a starting point?

I was thinking if we can execute `repo2docker` from the "outside" (with most of the other tests we only get to run code in `verify` inside the container) then we can use a [`start` file](https://repo2docker.readthedocs.io/en/latest/config_files.html#start-run-code-before-the-user-sessions-starts) that is executed inside the container to make a change and check that the local directory was indeed changed.

evertrol commented:
I've put the documentation as a FAQ (even if it isn't (yet)), since I'm now sure how well it fits as a how-to. And I didn't see documentation on the individual options (other than what argparse produces on the command line, obviously).

Also: I can't tick the check-box :-(

evertrol commented:
I seem to have hit a little snag: I can't bind an external volume to an internal volume using an internal variable. I'd like to bind to `$HOME` or `~`, but that just get escaped inside `client.containers.run(..., volumes=container_volumes, ...)`. I can get hardcode `/home/jovyan` for tests, but that is obviously not a solution when using user-supplied Docker files.

I've been looking at another way to work around this, or find a configuration option to change this, but no luck so far. Would anyone know how to do this?

(to be clear, I'd like to do:
```
client.containers.run(..., volumes={'/host/directory', {'bind': '$HOME', 'mode': 'rw'}}, ...)
```
)

betatim commented:
I think we could resolve `$HOME` before making that call?

I'd be Ok with having the docs say "Editable mode is a convenience option that will mount the repository to `/home/$USER`, if you need to mount to a different location in the container please use `--volumes`." if we can't find a solution to this.

betatim commented:
You could inspect `image['ContainerConfig']['Env']` to find `$HOME`. This would make it a bit harder to reason about which directory inside the container will be used as it would be "Relative to WorkingDir if set, otherwise relative to $HOME if set, if neither is set relative paths aren't supported"

I think it is fine to assume that `WorkingDir` is set and error out if not. The general philosophy for repo2docker is that if you use your own `Dockerfile` you are an expert and know what you are doing. We offer very little support to you when things don't work beyond "do you really need to make your own Dockerfile?". Conversely we try to support the things a lot of users want to do natively so that they do not have to write a `Dockerfile`.

So I would keep the current behaviour and add a test that a `Dockerfile` with WorkingDir set works as expected.

evertrol commented:
I've added two tests:

- a test that creates a file inside the container, for which it is checked that it can be viewed by the host (with its contents verified)
- a test that creates a file on the host, while the container is running. The existence of that file is then verified inside the container (though not its contents).

----

Additionally, I have converted the FAQ from Markdown to reStructuredText, since I wanted an admonition note, which is not possible with Markdown. Perhaps this change should be taken out of this PR and put into its own PR.

Lastly, I was looking for a CHANGES file to add this change, but there isn't any. Perhaps the repository is considered the CHANGES file, but generally, I find a separate file easier to read and spot (important) changes between versions, or bug-fixes.
Perhaps it's something to consider? Though I notice (sadly) few (notable) Python projects actually have such a file.

betatim commented:
Having a [ChangeLog would be welcome](https://github.com/jupyter/repo2docker/issues/376#issue-351079554) at least by me. We haven't been very disciplined so far so if you oulc get us started that would be great.

Do you have a preferred format? My suggestion would be to start with a style a bit like https://github.com/scikit-learn/scikit-learn/blob/master/doc/whats_new/v0.19.rst#version-0191.

Changing from MD to rst would be ideal in a separate PR so we can keep reformatting separate from content changes.

betatim commented:
I'll merge this now to get this cool feature in there.

neuromusic commented:
sorry for taking so long to say "thanks" but this is awesome! going to test it out ASAP
